components/datetime/src/format/datetime.rs

```diff
@@ -569,62 +554,107 @@ where
-                .map(FixedDecimal::from)
             {
-                Ok(x) => r = r.and(try_write_number(w, fixed_decimal_format, x, field.length)?),
+                Ok(x) => {
+                    r = r.and(try_write_number(
+                        w,
+                        fdf,
+                        usize::from(x).into(),
+                        field.length,
+                    )?)
+                }
                 Err(e) => {
                     r = r.and(Err(e));
-                    w.with_part(Part::ERROR, |w| "<ICU4X ERROR>".write_to(w))?
+                    w.with_part(Part::ERROR, |w| "{minute}".write_to(w))?
                 }
             };
         }
         FieldSymbol::Second(Second::Second) => {
             let seconds = datetime
                 .datetime()
                 .second()
-                .ok_or(Error::MissingInputField(Some("second")))
-                .map(usize::from)
-                .map(FixedDecimal::from)
-                .and_then(|mut seconds| {
+                .ok_or(Error::MissingInputField(Some("second")));
+
+            let () = match seconds {
+                Ok(s) => {
                     if let Some(PatternItem::Field(next_field)) = next_item {
                         if let FieldSymbol::Second(Second::FractionalSecond) = next_field.symbol {
-                            let mut fraction = FixedDecimal::from(usize::from(
-                                datetime
+                            // We only support fixed field length for fractional seconds.
+                            if let FieldLength::Fixed(p) = next_field.length {
+                                match datetime
                                     .datetime()
                                     .nanosecond()
-                                    .ok_or(Error::MissingInputField(Some("nanosecond")))?,
-                            ));
-
-                            // We only support fixed field length for fractional seconds.
-                            let precision = match next_field.length {
-                                FieldLength::Fixed(p) => p,
-                                _ => {
-                                    return Err(Error::Pattern(
-                                        crate::pattern::PatternError::FieldLengthInvalid(
-                                            FieldSymbol::Second(Second::FractionalSecond),
-                                        ),
-                                    ))
+                                    .ok_or(Error::MissingInputField(Some("nanosecond")))
+                                {
+                                    Ok(n) => {
+                                        r = r.and(try_write_number(
+                                            w,
+                                            fdf,
+                                            // s is integer, n * 10^-9 is fractional
+                                            #[allow(clippy::unwrap_used)]
```

Comment:
Issue: I don't think we enforce that `n` is less than 1e10, so this `unwrap` is reachable

Reply:
No we do, https://github.com/unicode-org/icu4x/blob/d09c8bd59a5f092e54e978ccaaad2bf30c2aa668/components/calendar/src/types.rs#L318

Reply:
Ok. That's fine, but we should still use `debug_assert`, since `unwrap` creates undesirable machine code.

Please switch this to using `concatenate_end`, and if it fails to concatenate for some reason, `debug_assert!` and throw away the fractional second digits. Note that you cannot do this with `concatenated_end`: https://github.com/unicode-org/icu4x/issues/4857

Reply:
> if it fails to concatenate for some reason

We're satisfying the precondition of `concatenate_end`, so it will not fail.